### PR TITLE
find remote reference when HEAD detached

### PR DESCRIPTION
**Pull request by gnalh**
Scan the references to find the matching reference name. This is equivalent to `git for-each-ref --contains e0c6eaa19870c1ae30601bed5bee604c7436c9a2`

After:
```
2024-04-08T05:37:42 [INFO] - Found git_sha: Some("e0c6eaa19870c1ae30601bed5bee604c7436c9a2")
2024-04-08T05:37:42 [INFO] - Found git_branch: Some("refs/remotes/pull/47/merge")
```

Before
```
2024-04-08T05:39:17 [INFO] - Found git_sha: Some("e0c6eaa19870c1ae30601bed5bee604c7436c9a2")
2024-04-08T05:39:17 [INFO] - Found git_branch: None
```

Repro:
```
git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +e0c6eaa19870c1ae30601bed5bee604c7436c9a2:refs/remotes/pull/47/merge
git checkout --progress --force refs/remotes/pull/47/merge
```

### CODE REVIEW

File: src/scanner.rs [hunk 1]
```diff
@@ -161,8 +162,25 @@ impl BundleRepo {
...
+            let git_head_id = git_head.id().unwrap();
+            let mut git_head_branch = git_head.referent_name().map(|s| s.as_bstr().to_string());
+            reference.remote_branches()?.for_each(|r| {
+                match r {
+                    Ok(r) => {
+                        let target = r.target();
+                        let id = target.try_id();
+                        if id.is_some() && id.unwrap().to_string() == git_head_id.to_string() {
+                            git_head_branch = Some(r.name().to_path().to_str().unwrap().to_string());
+                        };
```
Comment (gnalh) [inline src/scanner.rs:176]: I'm sure there's a more elegant solution here and could use some rust pointers.

File: src/scanner.rs [hunk 2]
```diff
@@ -161,8 +162,26 @@ impl BundleRepo {
...
+                let git_head_id = git_head.id().unwrap();
+                let reference = git_repo.references().unwrap();
+                reference.remote_branches()?.for_each(|r| match r {
+                    Ok(r) => {
+                        let target = r.target();
+                        let id = target.try_id();
+                        if id.is_some() && id.unwrap().to_string() == git_head_id.to_string() {
+                            git_head_branch =
+                                Some(r.name().to_path().to_str().unwrap().to_string());
+                        };
```
Comment (gnalh) [inline src/scanner.rs:177]: I'm sure there's a more elegant solution here. If there is, any pointers would be greatly appreciated.

File: src/scanner.rs [hunk 3]
```diff
@@ -161,8 +162,30 @@ impl BundleRepo {
...
+                    match r {
+                        Ok(r) => {
+                            let target = r.target();
+                            let id = target.try_id();
+                            if id.is_some()
+                                && git_head.id().is_some()
+                                && id.unwrap().to_string() == git_head.id().unwrap().to_string()
+                            {
+                                git_head_branch =
+                                    Some(r.name().to_path().to_str().unwrap().to_string());
```
Comment (riya-n) [inline src/scanner.rs:178]:
```suggestion
                                    r.name().to_path().to_str().map(|s| s.to_string());
```
which I believe would either give you `None` or `Some(s.to_string())`

File: src/scanner.rs [hunk 4]
```diff
@@ -161,8 +162,30 @@ impl BundleRepo {
...
+                                break;
+                            };
+                        }
+                        Err(e) => {
+                            log::warn!("unable to match reference {:?}", e);
```
Comment (riya-n) [inline src/scanner.rs:183]: Why do we need to warn?

**Review by riya-n — APPROVED** [review]
Looks good, but not sure we need to match or warn. Just a suggestion, but we could do something like this as well
```
if git_head_branch.is_none() {
  for rb in git_repo.references()?.remote_branches()? {
    git_head_branch = rb.map_or(None, |r| {
      let id = r.target().try_id().map(|id| id.to_string());
      let git_head_id = git_head.id().map(|id| id.to_string());
      if id.is_some() && git_head_id.is_some() && id == git_head_id {
        return r.name().to_path().to_str().map(|s| s.to_string());
      }
      None
    });
  }
}
```

File: src/scanner.rs (hunk 4 above)
Comment (gnalh) [reply src/scanner.rs:183]: We can lower it to info. I think right now we want to be verbose on logging to help with debugging.

File: src/scanner.rs (hunk 3 above)
Comment (gnalh) [reply src/scanner.rs:178]: Nice, that's handy.

File: src/scanner.rs [hunk 5]
```diff
@@ -161,8 +162,30 @@ impl BundleRepo {
...
+                                && git_head.id().is_some()
+                                && id.unwrap().to_string() == git_head.id().unwrap().to_string()
+                            {
+                                git_head_branch =
+                                    r.name().to_path().to_str().map(|s| s.to_string());
+                                break;
+                            };
+                        }
+                        Err(e) => {
+                            log::info!("Unexpected error when trying to find reference {:?}", e);
```
Comment (gnalh) [inline src/scanner.rs:183]:
```suggestion
                            log::debug!("Unexpected error when trying to find reference {:?}", e);
```

**Pull request merged**